trunner/runners/wrappers.py

```diff
@@ -0,0 +1,335 @@
+            self.args,
+            cwd=self.cwd,
+            encoding='ascii'
+        )
+
+        self.read_output()
+        self.proc.wait()
+        if self.proc.exitstatus != 0:
+            raise PsuError(' Loading plo using psu failed, psu exit status was not equal 0!')
+
+
+class Gdb(StandardWrapper):
+    """Wrapper for gdb-multiarch program"""
+
+    def __init__(self, target, file, script, cwd=None):
+        args = [f'{file}', '-x', f'{script}']
+        super().__init__(target, 'gdb-multiarch', args, cwd)
+        self.jlink_device = ''
+        if 'zynq7000' in target:
+            self.jlink_device = 'Zynq 7020'
+
+    def read_output(self):
+        if is_github_actions():
+            logging.info('::group::Run gdb\n')
+
+        # When using docker, the additional newline may be needed
+        idx = self.proc.expect_exact(["(gdb) ", "Type <RET> for more"])
+
+        if idx == 1:
+            logging.info(self.proc.before)
+            self.proc.send('\r\n')
+            self.proc.expect_exact("(gdb) ")
+
+        self.proc.send('c\r\n')
+
+        logging.info(f'{self.proc.before}\n')
+        self.proc.close()
+
+        if is_github_actions():
+            logging.info('::endgroup::\n')
+
+    def run(self):
+        with GdbServer(self.jlink_device) as gdbserv:
+            # Use pexpect.spawn to run a process as PTY, so it will flush on a new line
+            self.proc = pexpect.spawn(
+                self.progname,
+                self.args,
+                cwd=self.cwd,
+                encoding='ascii',
+                timeout=8
+            )
+
+            try:
+                self.read_output()
+            except pexpect.TIMEOUT:
+                error = 'Prompt not seen after executing loading script!'
+                raise GdbError(error, gdb_output=self.proc.before, gdbserv_output=gdbserv.output())
+
+            self.proc.close()
+
+    def close(self):
+        self.proc.close()
+        if self.proc.exitstatus != 0:
+            logging.error('gdb failed!\n')
+            raise GdbError('Loading plo using gdb failed, gdb exit status was not equal 0!')
+
+
+class Phoenixd(BackgroundWrapper):
+    """ Wrapper for phoenixd program"""
+
+    def __init__(
+        self,
+        target,
+        port,
+        dir='.',
+        cwd=None,
+    ):
+        if cwd is None:
+            cwd = boot_dir(target)
+        self.cwd = cwd
+        self.port = port
+        self.dir = dir
+        self.dispatcher_event = None
+        super().__init__()
+
+    def _reader(self):
+        """ This method is intended to be run as a separated thread.
+        It searches for a line stating that message dispatcher has started
+        and then browses phoenixd output that may be needed for proper working """
+
+        msg = f'Starting message dispatcher on [{self.port}]'
+
+        try:
+            self.proc.expect_exact(msg, timeout=4)
+        except (pexpect.EOF, pexpect.TIMEOUT):
+            return self
```

Comment:
Why we retun self here? Just `return`

Reply:
When we use sth like that: `with Phoenixd(self.target, self.phoenixd_port) as phd:` and then call `phd.output()` it's needed. That's why it was done same way before:

https://github.com/phoenix-rtos/phoenix-rtos-tests/blob/3860a610f7d98d241dc4540924d36fd53fe9c9e2/trunner/runners/common.py#L219

Reply:
I don't get it, you pointed to the `__enter__` function. The `_rader` method is just a function that is called as a new thread, its return value is ignored

Reply:
Oh, my mistake. I will change it to `return`.